### DISCUSSION TITLE
Channel Form and `data-ee-version=` Params

### DISCUSSION THREAD

**Discussion by eric-lamb62** [ExpressionEngine/ExpressionEngine#3722] [Q&A]
An interesting discussion happened in the EE Slack where it was mentioned how EE now "leaks" it being installed as well as the version within Channel Form. The concern is that, while security through obscurity isn't advised to pursue, it was a low hanging fruit previously within the EE ecosystem that did provide some coverage that is now lost. 
https://eecms.slack.com/archives/C04CUNNR9/p1692830093777979

Am curious about any insights into this? Are there any benefits that were gained through this? Is this a bug or an intentional shift into third party tooling? 

Anyone have any thoughts on this? 

---
EDIT:
here's the discussion content for better reference

*** Just asking, but in the past, since Ellis Lab one of the most sold feature about Expression Engine was security. I remember the Obama campaign website used it, so did the Trump one, and many governments sites. I know security by obscurity is not security but one of the main things Expression Engine had over any other CMS was that you could not actually get information about the system. Similar to how hosting companies hide the Apache web server version from their headers and any good administrator does not expose information version about servers softwares running for potential malicious users scanning for vulnerable sites, you could not find any information if a site was running EE. I remember malicious users tried to figure out if it was WordPress or Drupal and just gave up because an EE site was just standard HTML and nothing more. It was really fun to see those attacks trying to sniff information. But somehow that has changed in the past years. Is there any reason why EE sites are not spilling data all over the place. And not just that you are using Expression Engine but even the version which makes it very easy to scan the web for vulnerable Expression Engine sites. An example is the channel form, it even shows the version you are using:
data-ee-version="6.4.14"
Not it only it tells everyone you are using EE but even the version you are using.

*** I'm confused

*** EE isn't secure any longer?

*** where does EE tell the world it's EE? (legit asking, I didn't know it did that)

*** went through a few templates looking for data-ee-version= and can't find it; this on like a channel form situation or something?

*** oh geeze, yeah, you mentioned channel form smh

*** Its on the source code when you are using channel forms. I did not said its insecure, I just say now its its same as other CMS but a bit worse because it actually exposes the exact version you are running as well

*** that's fair :slightly_smiling_face: didn't mean to imply

*** You can just look at the source code of any page that has an EE form. Which malicious scanners will also scan on the web

*** oh sure

*** That allows you can to scan for vulnerable sites assuming you know an exploit for a specific version

*** maybe is a cost/benefit thing?

*** The reason some people still run on EE 2 is probably because nobody knows they are. Otherwise they would have been exploited already.

*** 'cause, has you mention, SBO isn't really a thing

*** well, if they're running php5 they have bigger things to worry about (I'd think)

*** What cost benefit? Every major and cloud provider hides version from their web servers well, Apache, Nginx, you name it

*** This is actually against PCI scanning tests as well

*** You are not supposed to expose server side software versions

*** oh shit, that right?

*** You don't expose your PHP version, or your MySQL version, or Web server, etc.

*** Exposing header versions of server software in HTTP responses or any other communications can pose security risks. Here's why you should consider not doing so:
1. **Information Leakage**: Displaying the version of your server software gives potential attackers specific details about your environment. This is a type of information leakage that can be avoided.
2. **Exploit Targeting**: If an attacker knows the specific version of a server software you are running, they can tailor their attacks to exploit known vulnerabilities of that particular version. This can greatly reduce the time and effort required for an attacker to exploit a system.
3. **Simplifies Reconnaissance**: Reconnaissance is the phase where attackers gather information before launching an attack. If server versions are available, it makes this phase much easier for the attacker.
4. **Perceived Negligence**: Exposing detailed server headers might suggest to attackers that the organization isn’t taking other necessary security precautions seriously. It could be perceived as a sign of lax security practices, making your server a more tempting target.
5. **Outdated Versions**: Sometimes, organizations are slow to update their server software. Exposing an outdated version not only shows vulnerabilities, but it also might suggest to attackers that the organization doesn't prioritize security patches.
6. **Reduces Security by Obscurity**: While security by obscurity (relying solely on keeping things secret) is not a robust defense strategy, adding layers of obscurity can be a part of a defense-in-depth approach. By not revealing server versions, you add one more hurdle for an attacker.
7. **Compliance and Regulations**: Some industry regulations and standards discourage or even prohibit the exposure of such information as it can be seen as a security risk. Failing to adhere can result in penalties or a failure to meet audit requirements.
8. **Competitive Intelligence**: While not a direct security concern, exposing server versions and technologies can provide competitors insights into the technology stack and choices made by an organization, which could be used for competitive advantage.
To address this issue, administrators can:
- Configure their servers to suppress or modify software version details in HTTP headers or other communication.
- Regularly review and update server configurations.
- Periodically perform penetration testing and vulnerability assessments to ensure that unnecessary information isn't being leaked.
In essence, while exposing server software versions might seem harmless, it provides additional information that, when combined with other pieces of data, can aid attackers in their campaigns. It's a simple and effective measure to conceal this information and thereby reduce the attack surface.

*** oh lawdy

*** guess the question is, "why did EE start exposing the version in channel form?"

*** The version I don't know, but the EE in the forms, I think starting at version 5.

*** FWIW, I hear ya on the above wall o' text, but it's still SBO which gets weird

*** It was not there before version 5.

*** interesting

*** is it ONLY channel form or you find it elsewhere? (just curious)

*** Not its not only exposing you are using Expression Engine which is not that bad, but the exact version you have installed.

*** I only found this on channel forms so far.

*** no, I get that, it's just not scary to me to expose that

*** But I can't be for sure, since I did not bothered me before. I already knew it was doing this since version 6, but it was surprising to see the exact version numbering now as well.

*** i completely missed this O.o

*** I recently upgraded to the latest release to its telling exactly the version to everyone:
data-ee-version="6.4.14"

*** still though, I go back to "why" and "what benefit was gained by it" 'cause this was clearly an intentional thing; curious on that.

*** Maybe a bug, I don't see any benefit at all to expose anything in the rendered templates, since internally EE would already know what version you are running in case it requires it for something

*** or maybe headless and/or third party applications

*** i don't do much in that world but know it exists (edited) 

*** again, ima wait to understand it before i worry (edited) 

*** (don't disagree with you though dude; just view the weight of it differently)

**Comment by litzinger** [comment]
I recall asking for the version number to be added to the DOM when inside of the CP. That way one could target different versions with add-ons CSS and what not. This definitely sounds like something that should not be exposed on the front-end.

**Comment by TomJaeger** [comment]
Channel form should not have EE version in it.

**Comment by intoeetive** [comment]
@litzinger is right, this exists so the add-on could target different styles against different CMS versions. Perhaps it should not be included by default.

I should also mention that using Channel Forms by itself if the software exposure at some level. It's loading (most of the times) some CSS and JS by looking at which the people could more or less precisely identify the CMS and add-on versions that are used

**Comment by eric-lamb62** [comment]
I think @intoeetive makes a good point on how, even without the `data-ee-version` param being there on a channel form, it'd still be trivial(ish) to determine an EE site programmatically (am thinking the `themes/ee` and `themes/user` combo spoils the mystery handily though there are also hidden form fields that compound this). 

So, in the interest of avoiding security wac-a-mole, I'm not really sold on this as a problem to be efforted. There's no winning in spending time and money avoiding one potentiality when there are others right around the corner for the questionable goal of security through obscurity. It'll be a never ending struggle with no winning in the end. 

Still, I'm on team "security's not a binary" which could certainly skew my perspective. 

Just my 2 cents after reading the above.
